src/PerformanceSensitiveAnalyzers/Core/ExplicitAllocationAnalyzer.cs

```diff
@@ -0,0 +1,109 @@
+            s_localizableObjectCreationRuleTitleAndMessage,
+            s_localizableObjectCreationRuleTitleAndMessage,
+            DiagnosticCategory.Performance,
+            DiagnosticSeverity.Info,
+            isEnabledByDefault: true);
+
+        internal static DiagnosticDescriptor AnonymousObjectCreationRule = new DiagnosticDescriptor(
+            AnonymousObjectCreationRuleId,
+            s_localizablAnonymousObjectCreationRuleTitleAndMessage,
+            s_localizablAnonymousObjectCreationRuleTitleAndMessage,
+            DiagnosticCategory.Performance,
+            DiagnosticSeverity.Info,
+            isEnabledByDefault: true,
+            helpLinkUri: "http://msdn.microsoft.com/en-us/library/bb397696.aspx");
+
+        internal static DiagnosticDescriptor LetCauseRule = new DiagnosticDescriptor(
+            LetCauseRuleId,
+            s_localizableLetCauseRuleTitleAndMessage,
+            s_localizableLetCauseRuleTitleAndMessage,
+            DiagnosticCategory.Performance,
+            DiagnosticSeverity.Info,
+            isEnabledByDefault: true);
+
+        private static readonly object[] EmptyMessageArgs = Array.Empty<object>();
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+            ArrayCreationRule,
+            ObjectCreationRule,
+            AnonymousObjectCreationRule,
+            LetCauseRule);
+
+        protected override ImmutableArray<OperationKind> Operations => ImmutableArray.Create(
+            OperationKind.ArrayCreation,
+            OperationKind.ObjectCreation,
+            OperationKind.AnonymousObjectCreation);
+
+        protected override void AnalyzeNode(OperationAnalysisContext context, in PerformanceSensitiveInfo info)
+        {
+            if (context.Operation is IArrayCreationOperation)
+            {
+                context.ReportDiagnostic(Diagnostic.Create(ArrayCreationRule, context.Operation.Syntax.GetLocation(), EmptyMessageArgs));
+                return;
+            }
+
+            if (context.Operation is IObjectCreationOperation)
+            {
+                if (context.Operation.Type.IsReferenceType)
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(ObjectCreationRule, context.Operation.Syntax.GetLocation(), EmptyMessageArgs));
+                    return;
+                }
+
+                if (context.Operation.Parent.GetType().Name == "LazyConversionOperation")
+                {
+                    // TODO: Should this not be under the TypeConverter analyzer?
+                    context.ReportDiagnostic(Diagnostic.Create(ObjectCreationRule, context.Operation.Syntax.GetLocation(), EmptyMessageArgs));
+                    return;
+                }
+            }
+
+            if (context.Operation is IAnonymousObjectCreationOperation)
+            {
+                if (context.Operation.Syntax.IsKind(CodeAnalysis.CSharp.SyntaxKind.LetClause))
```

Comment:
❓ This appears to be specific to C#. Do we have (or need) the VB equivalent?

Reply:
There does indeed seem to be one, used in the same way:
```
Dim discountedProducts = From prod In products
                         Let Discount = prod.UnitPrice * 0.1
                         Where Discount >= 50
                         Select prod.ProductName, prod.UnitPrice, Discount

I cannot see another obvious way to differentiate `let` from the normal anonymous object creation other than the syntax kind, so adding an additional check for `VisualBasic.SyntaxKind.LetClause` would probably fix this here. I have, however, not written a line of VB.NET so I am not comfortable setting up the unit tests for the VB.NET scenarios.

Reply:
I'm OK with leaving this as long as you create a follow-up bug detailing the test gap.